### PR TITLE
Laboratory work #4, Badasyan Alexandra - 19FPL1 

### PR DESCRIPTION
**Pull request by suralmasha**

### CODE REVIEW

File: shapes/square.py [hunk 1]
```diff
@@ -4,31 +4,12 @@
...
-        """
-        pass
-
-    def get_diagonal(self):
-        """
-        Returns the diagonal length of a square
-        :return int: the diagonal length of a square
-        """
-        pass
+        super().__init__(uid, length, length)
```
Comment (ValeriyaKuznetsova) [inline shapes/square.py:15]: Why can we not implement these methods in Square?

**Review by ValeriyaKuznetsova — COMMENTED** [review]
Looks great. I have one question

Write to me here or in Telegram

File: shapes/square.py (hunk 1 above)
Comment (suralmasha) [reply shapes/square.py:15]: 1) in terms of geometry: we don't need to implement them. A square is a rectangle, consequently, it repeats all rectangle methods.

2) in terms of programming: the class "Square" inherits from the class "Rectangle". This means that it already has all the methods written in the parent class.

**Comment by ValeriyaKuznetsova** [comment]
10

File: stack/stack.py [hunk 2]
```diff
@@ -12,34 +12,49 @@ class Stack:
     """
 
     def __init__(self, data: Iterable = None):
-        pass
+        if not isinstance(data, type(None)):
```
Comment (ValeriyaKuznetsova) [inline stack/stack.py:15]: any suggestions for a better check?

File: stack/stack.py [hunk 3]
```diff
@@ -12,34 +12,49 @@ class Stack:
...
+        return len(self.data)
 
     def empty(self) -> bool:
         """
         Return whether stack is empty or not
         :return: True if stack does not contain any elements
                  False if stack contains elements
         """
+
+        return bool(not self.data)
```
Comment (ValeriyaKuznetsova) [inline stack/stack.py:60]: a shorter way?

File: stack/stack.py [hunk 4]
```diff
@@ -12,34 +12,49 @@ class Stack:
     """
 
     def __init__(self, data: Iterable = None):
```
Comment (ValeriyaKuznetsova) [inline stack/stack.py:14]: can you give me 5 examples of accepted data, like a list – [1, 2, 3,4 ]. Explain why they're possible

**Review by ValeriyaKuznetsova — COMMENTED** [review]
One question and minor comments.
Now it's 9

File: stack/stack.py (hunk 4 above)
Comment (suralmasha) [reply stack/stack.py:14]: tuple - (1, 2, 3, ...)
dict - {a: 1, b: 2, c: 3, ...}. to iterate keys 
dict - {a: 1, b: 2, c: 3, ...}. to iterate values
string - 'abcd'
generators like range(3)

everywhere we can iterate values

File: stack/stack.py (hunk 2 above)
Comment (suralmasha) [reply stack/stack.py:15]: Does this mean shorter or more practical?
If shorter, it's possible to write in one line:
self. data = list (data) if not isinstance(data, type(None)) else []
If it's about more practical, I'll think about it a little more…

File: stack/stack.py (hunk 3 above)
Comment (suralmasha) [reply stack/stack.py:60]: It's 
return not self.data
!!!

**Comment by ValeriyaKuznetsova** [comment]
10

File: queue_/queue_.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+# pylint: disable=invalid-name
+class Queue_:
+    """
+    Queue Data Structure
+    """
+
+    def __init__(self, data: Iterable = (), max_size=10):
+        self.max_size = max_size
+        if len(data) <= max_size:
+            self.data = list(data)
```
Comment (ValeriyaKuznetsova) [inline queue_/queue_.py:19]: do we really need this check?

File: queue_/queue_.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        else:
+            self.data = list(data[:max_size])
+
+    def full(self):
+        """
+        Return whether queue_ is full or not
+        :return: True if queue_ does is full.
+        False if the queue_ can contain more elements
+        """
+        return bool(len(self.data) == self.max_size)
```
Comment (ValeriyaKuznetsova) [inline queue_/queue_.py:29]: shorter

File: queue_/queue_.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    def change_max_size(self, value: int):
+        """
+        Changes max_size of queue_
+        :param value: value to be max_size of queue_
+        """
+        if isinstance(value, int) and value > 0:
+            self.max_size = value
+        else:
```
Comment (ValeriyaKuznetsova) [inline queue_/queue_.py:38]: need `else`?

File: queue_/queue_.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            raise ValueError
+
+    def put(self, element):
+        """
+        Add the element ‘element’ at the end of queue_
+        :param element: element to add to queue_
+        """
+        if not self.full():
+            self.data.append(element)
+        else:
```
Comment (ValeriyaKuznetsova) [inline queue_/queue_.py:48]: same

File: queue_/queue_.py (hunk 7 above)
Comment (ValeriyaKuznetsova) [reply queue_/queue_.py:38]: better first to raise an error and then do what you want to do
just more readable, I think

File: queue_/queue_test.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+Tests for Queue class.
+"""
+
+import unittest
+
+from queue_.queue_ import Queue_
+
+
+class QueueTestCase(unittest.TestCase):
```
Comment (ValeriyaKuznetsova) [inline queue_/queue_test.py]: Good tests

**Review by ValeriyaKuznetsova — COMMENTED** [review]
Now it's 8

It's good! Just some minor things

File: queue_/queue_.py (hunk 5 above)
Comment (suralmasha) [reply queue_/queue_.py:19]: This is necessary to ensure that a queue is not bigger than the max size.  The user can enter more elements than it is possible.

File: queue_/queue_.py (hunk 6 above)
Comment (suralmasha) [reply queue_/queue_.py:29]: return len(self.data) == self.max_size

**Review by ValeriyaKuznetsova — COMMENTED** [review]
like for start.py and tests))
finish, please

**Pull request closed without merging**